src/LendController.sol

```diff
@@ -0,0 +1,156 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity 0.8.13;
+
+import {ERC721, ERC721TokenReceiver} from 'solmate/tokens/ERC721.sol';
+import {SafeTransferLib, ERC20} from "solmate/utils/SafeTransferLib.sol";
+
+import {INFTLoanFacilitator} from './interfaces/INFTLoanFacilitator.sol';
+
+struct LoanTerms {
+    bool allowBuyouts;
+    bool allowLoanAmountIncrease;
+    uint16 interestRate;
+    uint128 amount;
+    uint32 durationSeconds;
+    address loanAsset;
+}
+
+interface NFTLoanTermsSource {
+    function terms(uint256 tokenId, address nft, address loanAsset, bytes calldata data) external returns (LoanTerms memory);
+}
+
+contract LendController is ERC721TokenReceiver {
+    using SafeTransferLib for ERC20;
+    
+    struct LenderSpec {
+        address lender;
+        uint256 amount;
+        bool pull;
+    }
+
+    struct BorrowRequest {
+        LenderSpec[] lendersSpecs;
+        address termsSource;
+        LoanTerms terms;
+    }
+
+    uint256 ONE = 1e18;
+    mapping(address => mapping(address => bool)) public termsApprovals;
+    mapping(address => mapping(address => uint256)) public lenderBalance;
+    mapping(uint256 => mapping(address => uint256)) public lenderLoanShares;
+
+    INFTLoanFacilitator public loanFacilitator;
+
+    constructor(INFTLoanFacilitator _facilitator) {
+        loanFacilitator = _facilitator;
+    }
+
+    event SetTermsApproval(address indexed from, address indexed termsContract, bool approved);
+
+    function setTermsApproval(address termsContract, bool approved) external {
```

Comment:
we'll probably want to have an amount based allowance and decrementing against it as funds go out.